### ISSUE TITLE
Networks Learns Nothing!

### ISSUE THREAD

**Issue by Dola47** [jeshraghian/snntorch#33]
* snntorch version: 0.2.8
* Python version: 3.7.4
* Operating System: Windows 10

### Description
Hi,

I tried to use the snnTorch to do exactly as you are doing in Tutorial 3 (without spike_grad), and also in the upcoming Tutorial 4 (applied spike_grad to stein neuron)  for Spiking CNNs. Moreover, I also converted my dataset from static to spike version as in Tutorial 4 using rate encoding.

However, usually, I found that the network learns nothing and the training loss goes down and up slightly, but the output numbers and accuracy are completely unintuitive.

For example, I used a Conventional CNN on my Dataset and I got 85% accuracy, and I used it with the same architecture and hyperparameters and nothing is learned.

I will be glad if you can have some direct suggestions about what may be the problem and why the network is learning nothing in the Spiking domain?

I will present some of the code here.

Looking forward to your response.

```
# Network Architecture
num_inputs = 1*95*50
num_outputs = 8

# Training Parameters
batch_size=25

# Temporal Dynamics
num_steps = 25
alpha = 0.7
beta = 0.8

dtype = torch.float
device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
```
```
class Net(nn.Module):
    def __init__(self):
        super().__init__()

    # initialize layers
        self.conv1 = nn.Conv2d(in_channels=1, out_channels=16, kernel_size=5, padding=1, stride=1)
        self.lif1 = snn.Stein(alpha=alpha, beta=beta)
        self.conv2 = nn.Conv2d(in_channels=16, out_channels=32, kernel_size=5, padding=1, stride=1)
        self.lif2 = snn.Stein(alpha=alpha, beta=beta)
        self.conv3 = nn.Conv2d(in_channels=32, out_channels=64, kernel_size=5, padding=1, stride=1)
        self.lif3 = snn.Stein(alpha=alpha, beta=beta)
        self.fc1 = nn.Linear(64*10*4, 500)
        self.lif4 = snn.Stein(alpha=alpha, beta=beta)
        self.fc2 = nn.Linear(500, 8)
        self.lif5 = snn.Stein(alpha=alpha, beta=beta)
        
    def forward(self, x):
        # Initialize LIF state variables and spike output tensors
        spk1, syn1, mem1 = self.lif1.init_stein(batch_size, 16, 46, 24)
        spk2, syn2, mem2 = self.lif1.init_stein(batch_size, 32, 22, 11)
        spk3, syn3, mem3 = self.lif3.init_stein(batch_size, 64, 10, 4)
        spk4, syn4, mem4 = self.lif4.init_stein(batch_size, 500)
        spk5, syn5, mem5 = self.lif5.init_stein(batch_size, 8)

        

        spk5_rec = []
        mem5_rec = []

        for step in range(num_steps):
            cur1 = F.max_pool2d(self.conv1(x), 2)
            spk1, syn1, mem1 = self.lif1(cur1, syn1, mem1)
            cur2 = F.max_pool2d(self.conv2(spk1), 2)
            spk2, syn2, mem2 = self.lif2(cur2, syn2, mem2)
            cur3 = F.max_pool2d(self.conv3(spk2), 2)
            spk3, syn3, mem3 = self.lif3(cur3, syn3, syn3)
            cur4 = self.fc1(spk3.view(batch_size, -1))
            spk4, syn4, mem4 = self.lif4(cur4, syn4, mem4)
            cur5 = self.fc2(spk4)
            spk5, syn5, mem5 = self.lif5(cur5, syn5, mem5)
            
            spk5_rec.append(spk5)
            mem5_rec.append(mem5)

        return torch.stack(spk5_rec, dim=0), torch.stack(mem5_rec, dim=0)

net = Net().to(device)
print(net)
```
```
loss_hist = []
test_loss_hist = []
counter = 0

# Outer training loop
for epoch in range(40):
    minibatch_counter = 0
    train_batch = iter(training_generator)

    # Minibatch training loop
    for data_it, targets_it in train_batch:
        data_it = data_it.to(device)
        targets_it = targets_it.to(device)
        
        batch_size = data_it.shape[0]
        spk_rec, mem_rec = net(data_it.view(batch_size, 1, 95, 50))
        log_p_y = log_softmax_fn(mem_rec)
        loss_val = torch.zeros((1), dtype=dtype, device=device)

        # Sum loss over time steps: BPTT
        for step in range(num_steps):
            loss_val += loss_fn(log_p_y[step], targets_it)

        # Gradient calculation
        optimizer.zero_grad()
        loss_val.backward()

        # Weight Update
        optimizer.step()

        # Store loss history for future plotting
        loss_hist.append(loss_val.item())

        # Test set
        test_data = itertools.cycle(val_generator)
        testdata_it, testtargets_it = next(test_data)
        testdata_it = testdata_it.to(device)
        testtargets_it = testtargets_it.to(device)

        # Test set forward pass
        batch_size = testdata_it.shape[0]
        test_spk, test_mem = net(testdata_it.view(batch_size, 1, 95, 50))

        # Test set loss
        log_p_ytest = log_softmax_fn(test_mem)
        log_p_ytest = log_p_ytest.sum(dim=0)
        loss_val_test = loss_fn(log_p_ytest, testtargets_it)
        test_loss_hist.append(loss_val_test.item())

        # Print test/train loss/accuracy
        if counter % 50 == 0:
            train_printer()
        minibatch_counter += 1
        counter += 1

loss_hist_true_grad = loss_hist
test_loss_hist_true_grad = test_loss_hist
```
![image](https://user-images.githubusercontent.com/45386621/114377745-4b3b4d80-9b87-11eb-9605-abfcc9eca0f5.png)

**Comment by jeshraghian** [comment]
Hi @Dola47,

Tutorial 4 is not yet completed. Aiming to have it completed by the end of April. 

I had a quick look at your code. You mentioned that your input is a spiking dataset. However, I don't see your forward pass iterating through the time dimension of the data.

I.e., if x is spiking with the number of steps in the first dimension, the following line:

`cur1 = F.max_pool2d(self.conv1(x), 2)`

may need to be:

`cur1 = F.max_pool2d(self.conv1(x[step]), 2)`

In general, a lack of learning in SNNs where the same error does not exist in an equivalent ANN will often be due to either (i) a mistake made in how time is handled in the forward pass, or (ii) a fundamental issue with SNNs, e.g., they're known to struggle to learn in excessively deep networks where ANNs do not struggle. 

I'll shortly open up a user forum so that coding questions can be directed there rather than here in GitHub issues! That might be more productive. Will pass a link to you once that's done.

**Comment by Dola47** [comment]
Hi @jeshraghian ,

Maybe I was not that clear, I mean I tried once to submit my static data as an input to my SCNN, which is exactly the code above. And once more I converted my data to a spiking version using rate encoding, and I added the time steps as you exactly mentioned ( I have not attached the code for that).

In both cases, the network learned nothing. 

Let me raise some questions:

1. Do we have any specific way to specify the time domain dimension for example as in LSTM in snnTorch, because for example the dimension of 95 in my dataset may be considered as the time domain dimension.
2. Are you considering the network above as a deep network? Because from my point of view it is not that deep and I was expecting that the SNN will learn something?
3. Do you have any suggestions that I might try to in order to get the SNN to learn something (for ex: use the time to first spike instead of rate encoding, assign a time-domain dimension, reduce the size of the network)?

Looking forward to your response.

Thanks.

**Comment by Dola47** [comment]
@jeshraghian Just to note, I reduced the architecture of my network, and the network started to fit the data :)

**Comment by jeshraghian** [comment]
@Dola47 thanks for the update! 

Feel free to keep the issue open so I can address your questions once I have more free time.

Until then, I have found that conv nets are quite sensitive to alpha & beta (membrane decay parameters). You might consider doing a parameter sweep there. 

It's also possible to make alpha and beta learnable parameters too. I'll add this to the to-do list!

**Comment by Dola47** [comment]
@jeshraghian  Thanks a lot for your reply.

One more thing I noticed if I encoded my dataset with Latency encoding usually the Train Set Accuracy and Test Set Accuracy in every Minitbatch are the same, is that normal?

![image](https://user-images.githubusercontent.com/45386621/114554646-3df59000-9c67-11eb-898a-11764160f9bd.png)

**Comment by jeshraghian** [comment]
> Are you considering the network above as a deep network? Because from my point of view it is not that deep and I was expecting that the SNN will learn something?

In general, this might be considered a moderately sized network in SNNs.
Having said that, training SNNs using backprop is quite a challenge on even moderately sized architectures, and highly susceptible to vanishing gradients. 
Nonetheless, other researchers have successfully achieved it on comparable architectures.
Although not guaranteed to work / this is still a relatively open research question, consider the following:

(i) using a surrogate gradient which will make the gradient more continuous; 
(ii) sweeping across alpha/beta; 
(iii) increase the width of FCNs or depth of 2d convs which not only increase capacity, but would be expected to increase the amount of firing, and therefore, keep the gradient flowing.
(iv) modify weight initialization parameters to reduce vanishing gradients.

> Do you have any suggestions that I might try to in order to get the SNN to learn something (for ex: use the time to first spike instead of rate encoding, assign a time-domain dimension, reduce the size of the network)?

The tutorials show how your network can learn something using rate coding.
If you mean, how to implement a network that responds with a latency / time to first spike target, the the tutorials are enough to get you started with coding that together. Making the necessary tweaks would need a basic understanding of PyTorch.

From my own experience, cross-entropy loss in PyTorch will not natively support latency coding because an output target is expected at every step. I.e., one output classes must fire at every step. 
To get around this, you can either a) add a dummy neuron which fires when none of the true classes fire. More appropriately, I would b) apply a MSE Loss to the membrane potential, where the membrane is set to the threshold (or slightly above) at the desired firing time, t, and 0 elsewhere. 

There's no settled 'best' way just yet. As we discover best practices, we can add these to the examples section of the documentation.
But if you make any discoveries, keep us posted :-)

> One more thing I noticed if I encoded my dataset with Latency encoding usually the Train Set Accuracy and Test Set Accuracy in every Minitbatch are the same, is that normal?

I take this to mean your input uses latency coding, and your output targets are classified based on rate coding. 
The results do look quite odd; obviously make sure that you're not passing the same dataloader into your train and test sets. 
Otherwise, test the accuracy of the full training set to see if this issue persists. 

I'll close this issue for now, so please direct questions related to debugging training code, tutorials, and other general SNN advice to the GitHub Discussions forum. Hopefully this will encourage the broader community to assist. 
Issues can be left for errors arising from snnTorch itself. 

Thanks!

**Comment by Dola47** [comment]
@jeshraghian  Really thanks a lot for your response.

However, I started to fill I am missing some ideas:

I could not understand precisely what do you mean by that?

> From my own experience, cross-entropy loss in PyTorch will not natively support latency coding because an output target is expected at every step. I.e., one output classes must fire at every step.
To get around this, you can either a) add a dummy neuron which fires when none of the true classes fire. More appropriately, I would b) apply a MSE Loss to the membrane potential, where the membrane is set to the threshold (or slightly above) at the desired firing time, t, and 0 elsewhere.

Also here I want to note that I am not using the same dataloader. 

> I take this to mean your input uses latency coding, and your output targets are classified based on rate coding.
The results do look quite odd; obviously make sure that you're not passing the same dataloader into your train and test sets.
Otherwise, test the accuracy of the full training set to see if this issue persists.

Because both of the above points are quite related. I would like to sum up how I used the Latency code for my case. And I think the best thing is to present the code where the data is converted to spikes using Latency code.

Here is the code I used for the latency code:

```
loss_hist = []
val_loss_hist = []
counter = 0
#spike_data_list=[]
#spike_target_list=[]

# Outer training loop
for epoch in range(3):
    minibatch_counter = 0
    train_batch = iter(training_generator)

    # Minibatch training loop
    for data_it, targets_it in train_batch:
        data_it = data_it.to(device)
        targets_it = targets_it.to(device)
        batch_size = data_it.shape[0]
        # Spike generator
        spike_data, spike_targets = spikegen.latency(data_it, targets_it, num_outputs=num_outputs,
                                                     num_steps=num_steps, tau=5, threshold=0.01)
        #spike_data_list.append(spike_data)
        #spike_target_list.append(spike_targets)
        
        output, mem_rec = net(spike_data.view(num_steps, batch_size, 95, 32, 32))
        log_p_y = log_softmax_fn(mem_rec)
        loss_val = torch.zeros((1), dtype=dtype, device=device)

        # Sum loss over time steps to perform BPTT
        for step in range(num_steps):
            loss_val += loss_fn(log_p_y[step], targets_it)

        # Gradient calculation
        optimizer.zero_grad()
        loss_val.backward(retain_graph=True)
        nn.utils.clip_grad_norm_(net.parameters(), 1)
        
        # Weight Update
        optimizer.step()

        # Store loss history for future plotting
        loss_hist.append(loss_val.item())

        # Val set
        val_data = itertools.cycle(val_generator)
        valdata_it, valtargets_it = next(val_data)
        valdata_it = valdata_it.to(device)
        valtargets_it = valtargets_it.to(device)
        batch_size = valdata_it.shape[0]
        # Val set spike conversion
        val_spike_data, val_spike_targets = spikegen.latency(data_it, targets_it, num_outputs=num_outputs,
                                                             num_steps=num_steps, tau=5, threshold=0.01)
        
        # Val set forward pass
        val_output, val_mem_rec = net(val_spike_data.view(num_steps, batch_size, 95, 32, 32))

        # Val set loss
        log_p_yval = log_softmax_fn(val_mem_rec)
        log_p_yval = log_p_yval.sum(dim=0)
        loss_val_val = loss_fn(log_p_yval, valtargets_it)
        val_loss_hist.append(loss_val_val.item())

        # Print val/train loss/accuracy
        if counter % 50 == 0:
              train_printer()
        minibatch_counter += 1
        counter += 1

loss_hist_true_grad = loss_hist
val_loss_hist_true_grad = val_loss_hist
```

Simply it is the same as the one used for the rate encoding only I changed the functions.

It would be quite helpful to check If am doing it the right way to not, and if not. I will be very glad if you can suggest the modifications.

Thanks and looking forward to your response.

**Comment by Dola47** [comment]
Hi @jeshraghian ,

For your knowledge, I have been doing for a few days hyperparameter optimization on the alpha and beta parameters. I can say it is really used case-specific and every time I optimize I get different results. But, I can say normally alpha ranges from 0.4 up to 0.8. beta ranges from 0.3 to 0.9.

Moreover, you have a snippet of my code above that shows how I am using Temporal encoding and I am not sure if I am using it correctly or not. Please, if it is incorrect inform me. And I will be glad then if you can provide me with a few lines of code that show how to do it in the correct way.

Thanks a lot.

**Comment by jeshraghian** [comment]
Hi @Dola47 

> But, I can say normally alpha ranges from 0.4 up to 0.8. beta ranges from 0.3 to 0.9.

Thank you for the info! This is quite useful to know. 
I will soon run some tests that treat alpha and beta as learnable parameters.
My first tests where alpha and beta were applied globally to the whole network pushed alpha/beta either to 0 or >>1, and failed.
This might suggest alpha/beta must be unique to each layer, or unique to each neuron. 
This will take a while for me to get to while I focus on some paper deadlines, but I'll let you know how it goes once it is completed.

> Moreover, you have a snippet of my code above that shows how I am using Temporal encoding and I am not sure if I am using it correctly or not. Please, if it is incorrect inform me. And I will be glad then if you can provide me with a few lines of code that show how to do it in the correct way.

As I mentioned in the previous comment, could you please copy/paste your code-/theory-related questions in the [GH Discussions section](https://github.com/jeshraghian/snntorch/discussions/categories/q-a)? I'll gladly answer it there.
This can help other users who run into the same issues.
